### PR TITLE
Add support for Azure instances in testnet creation

### LINKED ISSUE #2028: Add Azure support to net/
Should just be a matter of writing a cloud provider plugin like the existing AWS and GCP providers

### PR DESCRIPTION
**Pull request by danpaul000**
#### Problem

Add support for Azure instances in testnet creation

#### Summary of Changes

Add API azure-provider.sh and a few other small changes to support the bringup of Azure instances as testnet nodes.

Fixes #2028

### CODE REVIEW

File: net/net.sh [hunk 1]
```diff
@@ -294,6 +294,7 @@ startNode() {
       "
   ) >> "$logFile" 2>&1 &
   declare pid=$!
+  #ln -sfT "fullnode-$ipAddress.log" "$netLogDir/fullnode-$pid.log"
```
Comment (mvines) [inline net/net.sh]: oh please remove this line

File: net/scripts/azure-provider.sh [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+#
+cloud_FetchFile() {
+  declare instanceName="$1"
+  declare publicIp="$2"
+  declare remoteFile="$3"
+  declare localFile="$4"
+
+  cloud_GetConfigValueFromInstanceName "$instanceName" osProfile.adminUsername
+  scp "${config_value}@${publicIp}:${remoteFile}" "$localFile"
+
```
Comment (mvines) [inline net/scripts/azure-provider.sh]: nit: remove stray whitespace

File: net/scripts/azure-provider.sh [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+      echo "Error: address may not be supplied when provisioning multiple nodes: $optionalAddress"
+      exit 1
+    }
+    args+=(
+      --public-ip-address "$optionalAddress"
+    )
+  fi
+
+  (
+  set -x
```
Comment (mvines) [inline net/scripts/azure-provider.sh]: fix indent please

File: net/scripts/azure-provider.sh [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+  declare enableGpu="$4"
+  declare machineType="$5"
+  declare zone="$6"
+  declare optionalBootDiskSize="$7"
+  declare optionalStartupScript="$8"
+  declare optionalAddress="$9"
+  declare optionalBootDiskType="${10}"
+
+  if $enableGpu; then
+    # TODO: First out what image is needed for Azure for GPU/non-GPU
```
Comment (mvines) [inline net/scripts/azure-provider.sh]: "First out" => "Figure out" ?

File: net/scripts/azure-provider.sh [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    imageName="UbuntuLTS"
+  else
+    # TODO: First out what image is needed for Azure for GPU/non-GPU
```
Comment (mvines) [inline net/scripts/azure-provider.sh]: Is this TODO correct?

File: net/scripts/azure-provider.sh [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+#  Examples of plain-text filter command
+#
+#  This will return an exact match for a machine named pgnode
+#  az vm list -d --query "[?name=='pgnode'].[name,publicIps,privateIps,location]"
+#
+#  This will return a match for any machine with prefix pgnode, ex: pgnode and pgnode2
+#  az vm list -d --query "[?starts_with(name,'pgnode')].[name,publicIps,privateIps,location]"
+
+__cloud_FindInstances() {
+
```
Comment (mvines) [inline net/scripts/azure-provider.sh]: nit: stray whitespace

File: net/scripts/azure-provider.sh (hunk 5 above)
Comment (danpaul000) [reply net/scripts/azure-provider.sh]: Yes, there is an extension needed to install the CUDA drivers.  I am working on figuring out the best way to add this in, or if I identify a custom image to take care of it upon VM creation.

Comment (mvines) [reply net/scripts/azure-provider.sh]: The TODO should only be in the `$enableGpu` case though right and not in the else (no GPU) block?

**Comment by danpaul000** [comment]
Yes, that's correct.  I'll remove it from both in my next commit likely
anyway.

**Comment by danpaul000** [comment]
Messed up my rebase.  Standby please.

File: net/net.sh [hunk 7]
```diff
@@ -294,7 +294,7 @@ startNode() {
       "
   ) >> "$logFile" 2>&1 &
   declare pid=$!
-  ln -sf "fullnode-$ipAddress.log" "$netLogDir/fullnode-$pid.log"
+  ln -sfT "fullnode-$ipAddress.log" "$netLogDir/fullnode-$pid.log"
```
Comment (danpaul000) [inline net/net.sh]: This was the original form, including the -T option.  I had commented it out previously when testing on Mac OSX, which doesn't support that option.

**Comment by garious** [comment]
:shipit:

**Pull request merged**